### ISSUE TITLE
snmp module doubling data

### ISSUE THREAD

**Issue by l2isbad** [firehol/netdata#981]
Hello. There is problem with collecting data using snmp module and incremental algorithm.

The problem is doubling data after retransmit request or service restart .

As an example:

update interval is 30 sec.

Request at 20:48:30 was droped.

Retransmit at 35
20:48:35.061532 IP y.y.y.y.33835 > z.z.z.z.161:  C=xxx GetRequest(84)  .1.3.6.1.2.1.31.1.1.1.6.1014 .1.3.6.1.2.1.31.1.1.1.10.1014 .1.3.6.1.2.1.31.1.1.1.6.19 .1.3.6.1.2.1.31.1.1.1.10.19

1st interval after retransmit is OK
20:49:00.028019 IP y.y.y.y.33835 > z.z.z.z.161:  C=xxx GetRequest(83)  .1.3.6.1.2.1.31.1.1.1.6.1014 .1.3.6.1.2.1.31.1.1.1.10.1014 .1.3.6.1.2.1.31.1.1.1.6.19 .1.3.6.1.2.1.31.1.1.1.10.19

2nd interval after retransmit is OK
20:49:30.085017 IP y.y.y.y.33835 > z.z.z.z.161:  C=xxx GetRequest(83)  .1.3.6.1.2.1.31.1.1.1.6.1014 .1.3.6.1.2.1.31.1.1.1.10.1014 .1.3.6.1.2.1.31.1.1.1.6.19 .1.3.6.1.2.1.31.1.1.1.10.19

3d interval after retransmit is OK
20:50:00.046215 IP y.y.y.y.33835 > z.z.z.z.161:  C=xxx GetRequest(83)  .1.3.6.1.2.1.31.1.1.1.6.1014 .1.3.6.1.2.1.31.1.1.1.10.1014 .1.3.6.1.2.1.31.1.1.1.6.19 .1.3.6.1.2.1.31.1.1.1.10.19

4d There we have doubling
20:50:30.099216 IP y.y.y.y.33835 > z.z.z.z.161:  C=x GetRequest(83)  .1.3.6.1.2.1.31.1.1.1.6.1014 .1.3.6.1.2.1.31.1.1.1.10.1014 .1.3.6.1.2.1.31.1.1.1.6.19 .1.3.6.1.2.1.31.1.1.1.10.19

So graph is looks like https://postimg.org/image/a6ttfvagp/

**Comment by ktsaou** [comment]
I am not exactly sure I get the issue.

When you say "doubling data" you means that netdata reports twice the value in the y-axis? At your chart, for example you believe that the spikes at 40 are wrong and the actual data said 20?

The SNMP plugin does not process the data. It just collects them and passes them to netdata. The incremental algorithm is implemented in netdata and is the same for all charts. This algorithm is tested with unit tests, so I think it is not buggy. The problem is somewhere else...

I tried to see this behaviour on my SNMP switch:

![image](https://cloud.githubusercontent.com/assets/2662304/18616539/60e4ea9e-7dc6-11e6-8f1d-5629b41f9ca3.png)

Note that I restarted netdata to follow the path you suggest.

Of course I will fix it if it is a netdata bug, but I need your help to figure out what is happening.

So, how have you verified this information is not sent by your switch?

**Comment by l2isbad** [comment]
Yes i am 100% sure. Numbers on my screen is Gb/s (gigabit per second) and there is no chance for 20-40-20 traffic patterns. Also im using tcpdump to collect snmp dump (-w name.pcap then basic math to count bandwidth per second to double check that is not problem with switch).

And it is very easy to reproduce. You should drop first Request packet with iptables rule (to simulate network drops) to force netdata send to send another Req (retransmit). And after 3-4 "update_every" i have value=not current but x2.

**Comment by ktsaou** [comment]
ok, let's find out...

Please run this:

```sh
# stop netdata - a few switches have issues with concurrent use on their SNMP module
killall netdata

# run the plugin in debug mode - it will print the values on screen
/usr/libexec/netdata/plugins.d/node.d.plugin debug 1 snmp
```

In the output, there is a probing phase where the plugin tries to find the available work to be done. At this stage it prints something like this:

```
Waiting modules configuration, still running 1
```

Then it starts polling the switch printing a lot of times `runAllServices()`.

At this stage, you will find 2 sets of values we are interested.

The following is data collected from the snmp device:

```
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.10.17 = 6966873
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.16.17 = 1002056817
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.10.18 = 3456176069
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.16.18 = 4222994086
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.10.19 = 3564472092
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.16.19 = 3670958612
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.10.20 = 75570953
Sun Sep 18 2016 19:09:53 GMT+0300 (EEST): node.d.plugin: DEBUG: snmp: 10.11.12.8: found snmp value of OIDs 1.3.6.1.2.1.2.2.1.16.20 = 1067610378
```

The following is data sent to netdata:

```
Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: sent chart snmp_switch.bandwidth_port15
Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: setting current chart to snmp_switch.bandwidth_port16
BEGIN snmp_switch.bandwidth_port16 13230000
SET in = 3810057967
SET out = 3137166791
END

Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: sent chart snmp_switch.bandwidth_port16
Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: setting current chart to snmp_switch.bandwidth_port17
BEGIN snmp_switch.bandwidth_port17 13230000
SET in = 6966873
SET out = 1002056817
END

Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: sent chart snmp_switch.bandwidth_port17
Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: setting current chart to snmp_switch.bandwidth_port18
BEGIN snmp_switch.bandwidth_port18 13230000
SET in = 3456176069
SET out = 4222994086
END

Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: sent chart snmp_switch.bandwidth_port18
Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: setting current chart to snmp_switch.bandwidth_port19
BEGIN snmp_switch.bandwidth_port19 13230000
SET in = 3564472092
SET out = 3670958612
END

Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: sent chart snmp_switch.bandwidth_port19
Sun Sep 18 2016 19:09:55 GMT+0300 (EEST): node.d.plugin: DEBUG: setting current chart to snmp_switch.bandwidth_port20
BEGIN snmp_switch.bandwidth_port20 13230000
SET in = 75570953
SET out = 1067610378
END
```

As you can see, the plugin is sending the values to netdata as-is. Could you please trap a section on your switch with these data?

Also, you can switch at your snmp configuration all values to `absolute`, so that we can compare them.

**Comment by l2isbad** [comment]
Hello.

- [ ] Start tcpdump (sudo tcpdump -n -i eth0 port 161 -l > test.pcap) to collect data (response time; response value)
- [ ] Start netdata.service

We are interesting in response only so

- [ ] awk '/Resp/ {print $1" "$8}' test.pcap | sed 's/.1.3.6.1.2.1.31.1.1.1.6.1014=//'        (format: time  SnmpResponseValue (64-bit version is ifHCInOctets))

[**1**] **_19:17:34_**.059512      95323278837022720.001713            [no data graph empty]
[**2**] **_19:18:00_**.128271      95323339587702784.003188            [no data graph empty]
[**3**] **_19:18:30_**.131021     95323410668744704.002223             [graph appears; 18,77 Gb/s; 19:18:30 time on graph]
[**4**] **_19:19:00_**.200333     95323482010710016.003775             [spike; 35,28 Gb/s; 19:19:00 time on graph]
[**5**] **_19:19:30_**.131515     95323553059696640.002916             [18,99 Gb/s; 19:19:30 time on graph]
[**6**] **_19:20:00_**.130984      95323624375603200.000719            [19,02 Gb/s; 19:20:00 time on graph]
[**7**] **_19:20:30_**.129715      95323695503212544.003923            [18,98 Gb/s; 19:20:30 time on graph]

Here iptables rule come and first packet at 19:21:00 was droped (next step is retransmit at 19:21:05)

[**8**] **_19:21:05_**.107716      95323775595606016.000087            [18,97 Gb/s; 19:21:05 time on graph]
[**9**] **_19:21:30_**.134045      95323837673783296.001202            [18,32 Gb/s; 19:21:35 time on graph]

After next step all graph time move +30 sec (update interval) - starting not at 19:18:30 but on 19:19:00 and so on)

[**10**] **_19:22:00_**.135495     95323909293940736.004093            [1st spike; 22,49 Gb/s; 19:22:00 time on graph]
[**11**] **_19:22:30_**.135048     95323980452085760.001138            [2nd spike; 35,18 Gb/s; 19:22:30 time on graph]
[**12**] **_19:23:00_**.129765     95324051725926400.002915            [OK; 19,01 Gb/s; 19:23:00 time on graph]

After all i have 9 steps on graph instead of 10 (19:19:00-19:23:00 intead of 19:18:30-19:23:00) and spikes. You can see that graph starts at 19:19:00 on my pic.

There is no step 7 (the one before Snmp GetReq was droped) on my graph (no 18,98 value; 19,02 then 18,97). 

Seems to fit 10 value graph to 9 value netdata summ step 7 value with 10 and 11. 

Calculation from snmp response (to double check spikes)

16.2001813504
18.954944512
19.0245240789
18.9463964373
19.0175750827
18.9673624917
21.3579715883
16.554180608
19.0987086507
18.975505344
19.006357504
![snmp-netdata](https://cloud.githubusercontent.com/assets/22274335/18629911/e4fd627e-7ea3-11e6-9c95-a6f6bf819ed4.jpeg)

**Comment by l2isbad** [comment]
**Same test with /usr/libexec/netdata/plugins.d/node.d.plugin debug 1 snmp**

- [ ] /opt/netdata/usr/libexec/netdata/plugins.d# ./node.d.plugin debug 1 snmp 1>/home/test/test.test 2>&1

1. Mon Sep 19 2016 20:37:12 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b30239e19f47**
2. Mon Sep 19 2016 20:37:30 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b30a6502769e**
3. Mon Sep 19 2016 20:38:00 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b31de8e39212**
4. Mon Sep 19 2016 20:38:30 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b3309fd5737a**
5. Mon Sep 19 2016 20:39:00 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b346c3e97402**
6. Mon Sep 19 2016 20:39:30 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b359dde39575**
7. Mon Sep 19 2016 20:40:00 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b369d46ae6c1**
8. Mon Sep 19 2016 **20:40:35** GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b3841ee0c947**
9. Mon Sep 19 2016 20:41:00 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b3907f4abc3d**
10. Mon Sep 19 2016 20:41:30 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b3a3956ed18a**
11. Mon Sep 19 2016 20:42:00 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b3b9e2286618**
12. Mon Sep 19 2016 20:42:30 GMT+0900 (YAKT): node.d.plugin: DEBUG: snmp: x.x.x.x: finished snmp with 1 successful and 0 failed values **SET in = 0x0152b3ca1375d285**

- [ ] awk '/SET/ {print $4}' test.test | xargs printf '%d \n' (convert to dec)

95335364270595911 
95335399353906846 
95335483170853394 
95335563549569914 
95335658644141058 
95335740684342645 
95335809244915393 
95335922163304775 
95335975320468541 
95336057296310666 
95336153072821784 
95336222619456133

Calculation

9.3555495808
22.3511857493
21.4343244032
25.3585523029
21.8773870891
18.2828194005
30.1115705003
14.1752436736
21.8602245632
25.5404029653
18.5457691605

**Comment by l2isbad** [comment]
Debug SET IN  vs tcpdump value  (same test)

**SET IN                                           Tcpdump**
```
95342893148800886           95342893148798976.001910
95342927188401490           95342927188398080.003410
95343012354528985           95343012354527232.001753
95343098939581524           95343098939580416.001108
95343184056488795           95343184056487936.000859
95343269212660760           95343269212659712.001048
95343354013664797           95343354013663232.001565
95343467036627194           95343467036626944.000250
95343524497875911           95343524497874944.000967
95343610321859946           95343610321858560.001386
95343695035942149           95343695035940864.001285
95343779709083345           95343779709079552.003793
```

**Calc**

```
9.0772268245            9.077226432
22.7109673344           22.7109677739
23.0893473451           23.0893475115
22.6978419413           22.6978420053
22.7083125248           22.7083124779
22.6136010752           22.6136009387
30.13945664             30.1394569899
15.3229996587           15.3229994667
22.8863957376           22.8863956309
22.5904219264           22.5904219477
22.5795043157           22.5795036501
```

All is ok

**Comment by ktsaou** [comment]
So, is netdata wrong?

**Comment by ktsaou** [comment]
I mean if you get unreasonably high values on the chart, we might have an issue in interpolation.

Btw, the first value in incremental charts is always lost. It is the base in calculations, since we try to find the incremental difference. Netdata also interpolates the values using microsecond calculations to predict the value that should be set at exactly second boundaries.

**Comment by ktsaou** [comment]
And for interpolation to work properly, the plugin is sending something like this:

```
BEGIN snmp_switch.bandwidth_port20 13230000
```

The number at the end, is the number of microseconds since the last data collection. Based on these numbers, netdata interpolates the values.

**Comment by l2isbad** [comment]
Yes it is. Netdata is very wrong.

Update interval: 30 sec

- **00:00:05** Start Netdata. 0st snmp request. After starting it is sending first SnmpRequest and value in response will be taken  into account on 3rd step

- **00:00:30** 1st snmp request. No data on graph.

- **00:01:00** 2nd snmp request. Graph appears and starts at 00:01:00. Dots on graph = 1. Should be = 1.

- **00:01:30** 3nd snmp request. Graph starts at 00:01:30. Here we have spike because of value on graph = current + (1snmp value - 0 snmp value). Dots on graph = 2. Should be = 2.

- **00:02:00** 4nd snmp request. Graph starts at 00:02:00. Value is ok.   Dots on graph = 3. Should be = 3.

**00:02:30** 5nd snmp request. Graph starts at 00:02:30. Value is ok.   Dots on graph = 4. Should be = 4.**

**so far all is ok besides 1 spike.**

Then we have 1 Req drop and retransmit after timeout (5 sec in my case)

- **00:03:05** 6nd snmp request. No new dots on graph. **Still 4. Should be = 5**. All graph moves on x-asis + 5 sec. So the first dot starts at 00:01:05

- **00:03:30**. 7nd snmp request. No new dots at all. **Still 4. Should be = 6**.

- **00:04:00** 8nd snmp request. 2 new dots on grap at once. Dots on graph = 6. Should be = 7. **All graph moves +1 update interval on x-asis!**. So one dot just disappear from graph.

- **00:04:30** 9nd snmp request. 1 new dot on graph. **Dots on graph = 7. Should be = 8**. Here i got big spike.

 - **00:04:00** 10nd snmp request. 1 new dot on graph. **Dots on graph = 8. Should be = 9**.


So after total 10 snmp req (actually 11 if we count the first one after start) i have completely wrong graph 

1. it moves + 1 update interval in time from start
2. it has 8 values instead of 9 (and +1 big spike)

The problem is very clear on high bandwidth  smooth traffic pattern interfaces.

The problem is how 
1. Netdata dealing with first snmp req after start service (taking it into account and using value to calculate).

**Same porblem with service restart (netdata sending packet and using its value). After restart in the middle of update interval after some iterations it moves begining of the graph on + update_interval sec! and removes one value from graph**

2. Netdata is dealing with changing update interval (30-30-35-25-30). It just removes one dot from graph and summ its value with next!

is it necessary to send first snmp packet not on update interval but on start/restart service? Should snmp module feed netdata with its value?

Also values from snmp module (BEGIN,SET) in debug mode are ok.

**Comment by ktsaou** [comment]
ok. There are 2 components involved:

- the snmp data collector.
- the netdata core that does all the calculations.

You haven't posted an entire session of the snmp data collector that reproduces the problem, so I can't tell where the problem is now.

You mention that the somehow netdata misses points on the charts. The timings data are collected are controlled by the snmp data collector and are sent to netdata like this:

```
BEGIN snmp_switch.bandwidth_port16 13230000
```

The last number is the number of microseconds (i.e. 1.000.000 = 1 second) since the last data collection sent to netdata for the same metric. A wrong calculation of this number can be responsible for both the missing points and the spikes.

The next possible problem is the interpolation netdata does. For example, we have a chart what appears a dot every 30 seconds. 00:00:00, 00:00:30, 00:01:00, etc. If for any reason a plugin sends values at 00:00:00, 00:00:35, 00:01:01, 00:01:33 etc netdata interpolates these values to find what they would have been at exactly the interpolation points the database needs.

If the problem is at the interpolation, we need to setup a unit test that reproduces. A unit-test is a scenario, with mockup data, like the ones found here: https://github.com/firehol/netdata/blob/ae37eb3bce368f5f1dacb9bd1d49137ab3ebde5b/src/unit_test.c#L583-L616

So, the first step is to verify that the last number of the `BEGIN` lines sent by the collector are right. It should be the number of microseconds since the same `BEGIN` line at the previous iteration. Can you check if this is right or wrong?

(I know this is painful, but I need your help to reproduce it. I you can't do this, please reproduce the problem once more and post the entire output of the snmp collector. I will do it.)

**Comment by l2isbad** [comment]
According my tests snmp data collector is working as intended -  all values are equal to tcpdump values and numbers of microsecodns are correct (30-30-30-35-25-30-30)

The problem occurr when netdata trying to normalize after  update interval shift - interpolation. (>=1 second delay is enough).

Raw dump output (exept runAllServices strings). http://pastebin.com/D0L4bwED

**Comment by ktsaou** [comment]
I verify this. It is a bug. I added a unit test and I am now working on it to fix it.

Thanks for reporting it!

**Comment by ktsaou** [comment]
Well, I fixed it! The code there is quite complicated, so it took me a few hours to figure it out. Without unit tests it is impossible to think all the cases.

Now you have your own unit test: https://github.com/ktsaou/netdata/blob/784ed60500457e249ed3893398430ce2246eb488/src/unit_test.c#L739-L805

It is in PR #1005. You can clone my repo to test it before merging it (https://github.com/ktsaou/netdata).

Thank you for reporting it and for providing all the information required!

**Comment by l2isbad** [comment]
Now its much better and there is no missing points and spikes but the problem is not completely solved - after restart service netdata is sending snmp request. 
Lets assume we have 30 sec update interval.
**t1**. 00:00:00
**t2**. 00:00:30
**t3** 00:01:00

**t4** then restart netdata at 00:01:29

**t5** and after regular update at 00:01:30 i have Value=t5-t4 (~0 because t5 value equal to t4)

i think there is no reason to send snmp req on start/restart service

**Comment by ktsaou** [comment]
I think that an underspike cannot be solved on netdata restart, without removing another point from the graph. Let's see why.

database time|collection time|collected value|incremental diff|on graph|
:--------------:|:---------------:|:------:|:------:|:-----:|
00:00:00|00:00:29|20|20|nothing - first point is discarded
00:00:30|00:00:30|21|1|1
00:01:00|00:01:00|41|20|20
00:01:30|00:01:30|61|20|20

**Comment by ktsaou** [comment]
I'll put this case in a unit test too. I think netdata attempts to extrapolate  the 1 second rate, to predict what would have been if it was like that for 30 seconds.

This is why, in some cases the 1st point may also show a small spike. netdata takes the rate of that one second and extrapolates it to find the rate of the whole period (30 seconds).

**Comment by ktsaou** [comment]
> i think there is no reason to send snmp req on start/restart service

**Comment by ktsaou** [comment]
wrong button... re-opened it

**Comment by l2isbad** [comment]
Hm just noticed. After last update something was broken.

update_interval 30

23:44:05 start service 0st poll
23:44:30 1st poll
23:45:00 NO dot only dimension highlighted 

since now graph always 
(update interval + (1st poll time - 0st poll time) behind  on mouseover 
(1st poll time - 0st poll time) behind with mouseout

and 1 dot is missing


since now graph always 
(update interval + (1st poll time - 0st poll time) behind  on mouseover 
(1st poll time - 0st poll time) behind with mouseout
And 1 dot is missing.

**Comment by ktsaou** [comment]
I think 0st and 1st poll are within 30 seconds.
I also found like that on the dump you gave above: https://github.com/ktsaou/netdata/blob/784ed60500457e249ed3893398430ce2246eb488/src/unit_test.c#L744 It is 13,5 seconds from the first.

Netdata creates a chart when the plugin says so, and keeps the update frequency. So, if you create a chart at 00:00:17 with update frequency 30, netdata will add values for 00:30:17, 01:00:17, 01:30:17, etc.

The node.js plugin however, tries to be aligned. So when you start it at 00:00:17 it will do one poll immediately, and the next will be 00:00:30, 00:01:00 (this is the first point on the chart), 00:01:30, etc.

So, it seems like a point is lost, but it is not. The plugin made 2 collections within the same interval.

I am adding a unittest for the case we discussed above.
